File: src/app_module.py
import os
from nest.common import Module
from injector import Injector, provider, singleton

from src.app_controller import AppController
from src.app_service import AppService
from src.data_feeds.base_feed import BaseDataFeed
from src.data_feeds.ccxt_provider_service import CcxtFeed
from src.data_feeds.fixed_feed import FixedFeed
from src.data_feeds.random_feed import RandomFeed


@singleton
@provider
def app_service_provider(injector: Injector) -> AppService:
    value_provider_impl = os.getenv("VALUE_PROVIDER_IMPL", "ccxt")

    if value_provider_impl == "fixed":
        data_feed = FixedFeed()
    elif value_provider_impl == "random":
        data_feed = RandomFeed()
    else:
        # Since CcxtFeed can be injected, we get it from the injector
        data_feed = injector.get(CcxtFeed)

    return AppService(data_feed=data_feed)


@Module(
    imports=[],
    controllers=[AppController],
    providers=[
        app_service_provider,
        CcxtFeed,
    ],
    exports=[app_service_provider, CcxtFeed],
)
class AppModule:
    pass

File: src/app.module.ts
import { Module } from "@nestjs/common";
import { ExampleProviderService } from "./app.service";
import { ExampleProviderController } from "./app.controller";
import { CcxtFeed } from "./data-feeds/ccxt-provider-service";
import { RandomFeed } from "./data-feeds/random-feed";
import { BaseDataFeed } from "./data-feeds/base-feed";
import { FixedFeed } from "./data-feeds/fixed-feed";

@Module({
  imports: [],
  controllers: [ExampleProviderController],
  providers: [
    {
      provide: "EXAMPLE_PROVIDER_SERVICE",
      // The useFactory provider dynamically selects the data feed implementation
      // based on the `VALUE_PROVIDER_IMPL` environment variable.
      useFactory: async () => {
        let dataFeed: BaseDataFeed;

        // If VALUE_PROVIDER_IMPL is set to "fixed", use the FixedFeed provider.
        if (process.env.VALUE_PROVIDER_IMPL == "fixed") {
          dataFeed = new FixedFeed();
          // If VALUE_PROVIDER_IMPL is set to "random", use the RandomFeed provider.
        } else if (process.env.VALUE_PROVIDER_IMPL == "random") {
          dataFeed = new RandomFeed();
          // Otherwise, default to the CcxtFeed provider.
        } else {
          const ccxtFeed = new CcxtFeed();
          await ccxtFeed.start();
          dataFeed = ccxtFeed;
        }

        // Instantiate the ExampleProviderService with the selected data feed.
        const service = new ExampleProviderService(dataFeed);
        return service;
      },
    },
  ],
})
export class AppModule {}
